### DISCUSSION TITLE
Direct SSH connection to a registered OpenSSH (agentless) node

### DISCUSSION THREAD

**Discussion by matjazp** [gravitational/teleport#26129] [Q&A] [answered]
SSH connection to a new registered (agentless) OpenSSH nodes via Teleport proxy works (GUI and CLI access). 

Is it possible to use this feature to connect **directly** to the host (so NOT via Teleport proxy?). I have a lot of globally accessible hosts with external IPs, but would still want to retain Teleport to control access to them via issuing SSH certificates. In this case it makes no sense to connect via proxy. Session logging probably won't work, but that's ok in those cases.

**Comment by webvictim** [comment]
I think this should be possible. After `tsh login`, try `ssh -o UserKnownHostsFile=~/.tsh/known_hosts login@external-ip`?

You might need to make a copy of `~/.tsh/known_hosts` to edit the list of allowed hostnames to include your SSH external IP, then pass that config file to `ssh` instead.

**Reply by matjazp** [reply]
I haven't been able to configure this for direct SSH access to the node. UserKnownHostsFile actually works, but I can't connect to the host directly, as sshd on the node rejects the certificate, signed by Teleport CA.

Looks like this is by design?

https://github.com/gravitational/teleport/blob/80f74172d3d376d7001b53cfe3ab843e2578a9ef/rfd/0098-registered-openssh-nodes.md?plain=1#L85

**Reply by webvictim** [reply]
Can you share the logs you get from `ssh -vv` when you're trying to connect? The logs from `sshd` on the server side would also be helpful.

It may be that the "new" way of joining SSH nodes (`teleport join openssh`) doesn't make this possible. The old way (`tctl auth sign --format=openssh --host=hostname`) absolutely does allow this though. You can see details of this in the older v12 docs: https://goteleport.com/docs/ver/12.x/server-access/guides/openssh/

**Reply by matjazp** [reply, marked as answer]
> It may be that the "new" way of joining SSH nodes (`teleport join openssh`) doesn't make this possible. The old way (`tctl auth sign --format=openssh --host=hostname`) absolutely does allow this though. You can see details of this in the older v12 docs: https://goteleport.com/docs/ver/12.x/server-access/guides/openssh/

Yes, you're right. `teleport join openssh` adds to TrustedUserCAKeys the Teleport's OpenSSH CA
`curl 'https://proxy/webapi/auth/export?type=openssh' | sed "s/cert-authority\ //"`, the "old manual" way adds normal
`curl 'https://proxy/webapi/auth/export?type=user' | sed "s/cert-authority\ //"` Teleport User CA.

If I replace OpenSSH CA with User CA, then I can ssh directly into the node (if there is an existing account on the node for username in SSH certificate), but then SSH access via Web UI or `tsh ssh` doesn't work anymore. But if I add both CA certificates to the `/var/lib/teleport/openssh/teleport_openssh_ca.pub` on the node, then both ways work - via proxy and directly.

This is a nice workaround, if you take extra care to create/delete user accounts on all directly accessible nodes, as direct SSH access using  User CA certs overrides all roles defined in Teleport.
